components/iam/pkg/oidc/service_test.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package oidc
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/oauth2"
+)
+
+func TestGetStartParams(t *testing.T) {
+	const (
+		issuerG  = "https://accounts.google.com"
+		clientID = "client-id-123"
+	)
+	service := NewOIDCService()
+	config := &OIDCClientConfig{
+		ID:         "google-1",
+		Issuer:     issuerG,
+		OIDCConfig: &oidc.Config{},
+		OAuth2Config: &oauth2.Config{
+			ClientID: clientID,
+			Endpoint: oauth2.Endpoint{
+				AuthURL: issuerG + "/o/oauth2/v2/auth",
+			},
+		},
+	}
+
+	params, err := service.GetStartParams(config)
+
+	require.NoError(t, err)
+	require.NotNil(t, params.Nonce)
+	require.NotNil(t, params.State)
+
+	// AuthCodeURL example:
+	// https://accounts.google.com/o/oauth2/v2/auth
+	// ?client_id=client-id-123
+	// &nonce=UFTMxxUtc5jVZbp2a2R9XEoRwpfzs-04FcmVQ-HdCsw
+	// &response_type=code
+	// &state=Q4XzRcdo4jtOYeRbF17T9LHHwX-4HacT1_5pZH8mXLI
+	require.NotNil(t, params.AuthCodeURL)
+	require.Contains(t, params.AuthCodeURL, issuerG)
+	require.Contains(t, params.AuthCodeURL, clientID)
+	require.Contains(t, params.AuthCodeURL, params.Nonce)
+	require.Contains(t, params.AuthCodeURL, params.State)
+}
+
+func TestGetClientConfigFromRequest(t *testing.T) {
+	issuer, err := setupFakeIdP(t)
+	require.NoError(t, err)
+
+	testCases := []struct {
+		Location      string
+		ExpectedError bool
+		ExpectedId    string
+	}{
+		{
+			Location:      "/start?word=abc",
+			ExpectedError: true,
+			ExpectedId:    "",
+		},
+		{
+			Location:      "/start?issuer=" + url.QueryEscape(issuer),
+			ExpectedError: false,
+			ExpectedId:    "google-1",
+		},
+		{
+			Location:      "/start?issuer=UNKNOWN",
+			ExpectedError: true,
+			ExpectedId:    "",
+		},
+	}
+
+	service := NewOIDCService()
+	err = service.AddClientConfig(&OIDCClientConfig{
+		ID:           "google-1",
+		Issuer:       issuer,
+		OIDCConfig:   &oidc.Config{},
+		OAuth2Config: &oauth2.Config{},
+	})
+	require.NoError(t, err, "failed to initialize test")
+
+	for _, tc := range testCases {
+		t.Run(tc.Location, func(t *testing.T) {
+			request := httptest.NewRequest(http.MethodGet, tc.Location, nil)
+			config, err := service.GetClientConfigFromRequest(request)
+			if tc.ExpectedError == true {
+				require.Error(t, err)
+			}
+			if tc.ExpectedError != true {
+				require.NoError(t, err)
+				require.NotNil(t, config)
+				require.Equal(t, tc.ExpectedId, config.ID)
+			}
+		})
```

Comment:
We should user `httptest.NewServer()` and bind the server handler, then invoke it with a request. That ensures we test including the whole stack (including the middlewares) rather than just the handler in isolation.

Reply:
Here it's a simple unit test for a single service operation which needs to inspect the http request. 

In other cases 👍🏻 
'Doing ping pong in https://github.com/gitpod-io/gitpod/pull/15064/files#diff-4fbad7c4fe24e0a9d817c00390cc699978f439366c58f67991e948900509006eR85

Reply:
here we go, commit https://github.com/gitpod-io/gitpod/pull/15440/commits/c2f516b2cf9979f41b928c040182c3381a7906ad
from https://github.com/gitpod-io/gitpod/pull/15440/commits